api/v1beta1/keystoneapi_types.go

```diff
@@ -128,19 +128,19 @@ type KeystoneAPISpec struct {
 
 	// +kubebuilder:validation:Optional
 	// NetworkAttachments is a list of NetworkAttachment resource names to expose the services to the given network
-	NetworkAttachments []string `json:"networkAttachments"`
+	NetworkAttachments []string `json:"networkAttachments,omitempty"`
 
 	// +kubebuilder:validation:Optional
 	// ExternalEndpoints, expose a VIP using a pre-created IPAddressPool
-	ExternalEndpoints []MetalLBConfig `json:"externalEndpoints"`
+	ExternalEndpoints []MetalLBConfig `json:"externalEndpoints,omitempty"`
 }
 
 // MetalLBConfig to configure the MetalLB loadbalancer service
 type MetalLBConfig struct {
 	// +kubebuilder:validation:Optional
 	// +kubebuilder:validation:Enum=internal;public
 	// Endpoint, OpenStack endpoint this service maps to
-	Endpoint endpoint.Endpoint `json:"endpoint"`
+	Endpoint endpoint.Endpoint `json:"endpoint,omitempty"`
```

Comment:
This is a strange one.  Optional would allow empty value which means "" for the golang string behind Endpoint, but the Enum validation rejects that value. So this is not really an optional field. I suggest to make this Required.

Reply:
for keystone itself yes, but when we move this struct to the infra-operator later, its optional with the intention that you could expose a service via a vip which is not used/registered as an endpoint

Reply:
What will be the value of the Endpoint golang struct field in that case? The current definition only allow two possible values `internal` and `public`.

Reply:
e.g. if the DB is ever needed to be exposed

Reply:
> What will be the value of the Endpoint golang struct field in that case? The current definition only allow two possible values `internal` and `public`.

you don't set an endpoint in that case where you don't register the vip as an endpoint. the alternative is we introduce a `none` value.

Reply:
Changing the validation to `// +kubebuilder:validation:Enum=internal;public;none`  would allow expressing if no Endpoint is needed. Still in that case I would make the defintion `// +kubebuilder:validation:Required` to avoid confusion as it always need one of the 3 possible values.

Reply:
right, if we introduce `none`, it can be required

Reply:
I'd suggest that we make this required and keep the current two values. Right now there is no need for the use case I described. We'll introduce that as a follow up as it also requires some work in lib-common.

Reply:
The whole optional, non defaulted CRD field + golang struct unmarshalling is problematic. Golang defines its own set of default values if the field is uninitialized. In this case the string endpoint field will be initialized to the "" string if the json request does not contain the field. But the marshaling of the golang struct back to json will pick up the """ string as a value for the json representation instead of skipping the field in the json output. This is where omitempty cames to the picture instructing the json emitter to skip the whole field if the value of the field happen to be the empty golang default value.

We saw in the past that using omitempty in a Spec field is problematic as it prevents forcing a the field value to the golang default if there is another value set before. See https://github.com/openstack-k8s-operators/nova-operator/pull/177

Reply:
Using defaults would probably work from serialization/deserialization pov for arrays, maps and struct and even fields, but, I think this has to change across all operator CRDs. We currently use omitempty everywhere including string fields[1] that are optional.

I think the current assumption is if a field is optional it can be omitted. IMO, we should not change that for a few newly added fields only.

[1] https://github.com/openstack-k8s-operators/keystone-operator/blob/master/api/v1beta1/keystoneapi_types.go#L173
https://github.com/openstack-k8s-operators/keystone-operator/blob/master/api/v1beta1/keystoneapi_types.go#L122

Reply:
I also feel that making it optional and providing a empty default is kind of counter intuitive.

Reply:
However, I'll make the change suggested to get me unblocked atm. But we probably need to be consistent across all CRDs and have some of these assumptions documented.